### PR TITLE
two commits about pci passthrough

### PR DESCRIPTION
**Pull request by PandaWei**
1) libvirt_xml/nodedev_xml: Add virt_functions in PCIXML class.
    
    This function had been reverted to fix conflicts.
    But it's useful and needed by pci-passthrough cases.
    so i modified it and pushed it again.
    Refer to commit e1152a766ef46e092634d8 for more details.

2)utils_test/libvirt: using try/except to deal with KeyError

### CODE REVIEW

File: virttest/libvirt_xml/nodedev_xml.py [hunk 1]
```diff
@@ -139,9 +154,56 @@ def __init__(self, virsh_instance=base.virsh):
...
+            accessors.XMLAttribute('slot', self, parent_xpath='/',
+                                   tag_name='address', attribute='slot')
+            accessors.XMLAttribute('function', self, parent_xpath='/',
+                                   tag_name='address', attribute='function')
+            super(PCIXML.Address, self).__init__(virsh_instance=virsh_instance)
+            self.xml = ('<address/>')
+
+    @staticmethod
+    def marshal_from_address(item, index, libvirtxml):
+        """Convert a Address instance into tag + attributes"""
```
Comment (will-Do) [inline virttest/libvirt_xml/nodedev_xml.py]: s/a Address/an address

File: virttest/libvirt_xml/nodedev_xml.py [hunk 2]
```diff
@@ -139,9 +154,56 @@ def __init__(self, virsh_instance=base.virsh):
...
+        root = item.xmltreefile.getroot()
+        if root.tag == 'address':
+            return (root.tag, dict(root.items()))
+        else:
+            raise xcepts.LibvirtXMLError("Expected a list of address "
+                                         "instances, not a %s" % str(item))
+
+    @staticmethod
+    def marshal_to_address(tag, attr_dict, index, libvirtxml):
+        """Convert a tag + attributes into a Address instance"""
```
Comment (will-Do) [inline virttest/libvirt_xml/nodedev_xml.py]: s/a Address/an Address

File: virttest/utils_test/libvirt.py [hunk 3]
```diff
@@ -691,15 +691,13 @@ def pci_label_from_address(address_dict, radix=10):
...
-    domain = int(address_dict['domain'], radix)
-    bus = int(address_dict['bus'], radix)
-    slot = int(address_dict['slot'], radix)
-    function = int(address_dict['function'], radix)
+    try:
+        domain = int(address_dict['domain'], radix)
+        bus = int(address_dict['bus'], radix)
+        slot = int(address_dict['slot'], radix)
+        function = int(address_dict['function'], radix)
+    except KeyError, detail:
```
Comment (will-Do) [inline virttest/utils_test/libvirt.py]: Good, and it's better to catch ```TypeError``` when using ```int()```

**Comment by PandaWei** [comment]
@will-Do
Thanks for your suggestion.
Updated them.

**Comment by will-Do** [comment]
OK

**Comment by PandaWei** [comment]
@will-do
can you help you to merge it if you have any other comment?
This PR has been going on for a long time

**Comment by will-Do** [comment]
Sure

**Pull request merged**